### PR TITLE
zenfs: add a new API for disk usage info

### PR DESCRIPTION
**Pull request by royguo**
Some databases would like to know the ZNS device's disk usage, so we should provide one.

Signed-off-by: Kuankuan Guo <guokuankuan@bytedance.com>

### CODE REVIEW

File: env/env_zenfs.cc [hunk 1]
```diff
@@ -485,7 +485,16 @@ class ZenfsEnv : public EnvWrapper {
 
   Status GetZbdDiskSpaceInfo(uint64_t& total_size, uint64_t& avail_size,
                              uint64_t& used_size) {
-    return Status::NotSupported("GetZbdDiskSpaceInfo is not implemented.");
+    auto zen_fs = dynamic_cast<ZenFS*>(fs_);
```
Comment (mm304321141) [inline env/env_zenfs.cc:488]: assert(zen_fs != nulltr);
if (zen_fs == nullptr) {
  // ???
}

File: include/rocksdb/env.h [hunk 2]
```diff
@@ -1681,4 +1681,8 @@ Status NewZenfsEnv(
     Env** zenfs_env, const std::string& zdb_path, std::string bytedance_tags_,
     std::shared_ptr<MetricsReporterFactory> metrics_reporter_factory_);
 
+#ifdef WITH_ZENFS
+Status GetZbdDiskSpaceInfo(Env* env, uint64_t& total, uint64_t& free, uint64_t& used);
```
Comment (mm304321141) [inline include/rocksdb/env.h]: 作为输出参数，total, free, used 都用指针，不要用引用（引用通常我们会认为该参数同时作为输入+输出使用）

File: env/env_zenfs.cc [hunk 3]
```diff
@@ -534,6 +543,16 @@ Status NewZenfsEnv(
   return s;
 }
 
+Status GetZbdDiskSpaceInfo(Env* env, uint64_t& total_size, uint64_t& avail_size,
+                           uint64_t& used_size) {
+  auto s = Status::OK();
+  auto zen_env = dynamic_cast<ZenfsEnv*>(env);
+  if (zen_env) {
+    s = zen_env->GetZbdDiskSpaceInfo(total_size, avail_size, used_size);
+  }
```
Comment (mm304321141) [inline env/env_zenfs.cc:571]: else {
s = NotSupported
}

File: include/rocksdb/env.h (hunk 2 above)
Comment (royguo) [reply include/rocksdb/env.h]: > 作为输出参数，total, free, used 都用指针，不要用引用（引用通常我们会认为该参数同时作为输入+输出使用）

Make sense

File: env/env_zenfs.cc (hunk 1 above)
Comment (royguo) [reply env/env_zenfs.cc:488]: Fixed

File: env/env_zenfs.cc (hunk 3 above)
Comment (royguo) [reply env/env_zenfs.cc:571]: Fixed

File: env/env_zenfs.cc [hunk 4]
```diff
@@ -534,6 +548,18 @@ Status NewZenfsEnv(
...
 }
 
+Status GetZbdDiskSpaceInfo(Env* env, uint64_t* total_size, uint64_t* avail_size,
+                           uint64_t* used_size) {
+  auto s = Status::OK();
+  auto zen_env = dynamic_cast<ZenfsEnv*>(env);
+  if (zen_env) {
+    s = zen_env->GetZbdDiskSpaceInfo(total_size, avail_size, used_size);
+  } else {
+    return Status::NotSupported();
```
Comment (mm304321141) [inline env/env_zenfs.cc:570]: *total_size = 0
*avail_size = 0
*used_size = 0

File: env/env_zenfs.cc [hunk 5]
```diff
@@ -534,6 +548,18 @@ Status NewZenfsEnv(
   return s;
 }
 
+Status GetZbdDiskSpaceInfo(Env* env, uint64_t* total_size, uint64_t* avail_size,
+                           uint64_t* used_size) {
+  auto s = Status::OK();
```
Comment (mm304321141) [inline env/env_zenfs.cc:562]: assert(total_size);
assert(avail_size);
assert(used_size);

File: env/env_zenfs.cc [hunk 6]
```diff
@@ -483,9 +483,23 @@ class ZenfsEnv : public EnvWrapper {
...
 
-  Status GetZbdDiskSpaceInfo(uint64_t& total_size, uint64_t& avail_size,
-                             uint64_t& used_size) {
-    return Status::NotSupported("GetZbdDiskSpaceInfo is not implemented.");
+  Status GetZbdDiskSpaceInfo(uint64_t* total_size, uint64_t* avail_size,
+                             uint64_t* used_size) {
+    auto zen_fs = dynamic_cast<ZenFS*>(fs_);
+    assert(zen_fs != nullptr);
+    if(zen_fs == nullptr) {
+        return Status::InvalidArgument("ZenFS: fs_ is empty!");
```
Comment (mm304321141) [inline env/env_zenfs.cc]: *total_size = 0
*avail_size = 0
*used_size = 0

File: env/env_zenfs.cc [hunk 7]
```diff
@@ -483,9 +483,23 @@ class ZenfsEnv : public EnvWrapper {
     target_->SanitizeEnvOptions(env_opts);
   }
 
-  Status GetZbdDiskSpaceInfo(uint64_t& total_size, uint64_t& avail_size,
-                             uint64_t& used_size) {
-    return Status::NotSupported("GetZbdDiskSpaceInfo is not implemented.");
+  Status GetZbdDiskSpaceInfo(uint64_t* total_size, uint64_t* avail_size,
+                             uint64_t* used_size) {
+    auto zen_fs = dynamic_cast<ZenFS*>(fs_);
```
Comment (mm304321141) [inline env/env_zenfs.cc:488]: assert(total_size);
assert(avail_size);
assert(used_size);

**Comment by mm304321141** [comment]
LGTM

**Pull request merged**